File: flutter_app/lib/core/offline/offline_queue.dart
import 'dart:convert';

import 'package:drift/drift.dart';
import 'package:uuid/uuid.dart';

import '../api/api_client.dart';
import 'offline_queue_db.dart';

enum SyncStatus { queuedLocal, syncing, synced, conflict, failed }

SyncStatus _statusFromString(String s) => SyncStatus.values.firstWhere(
  (v) => v.name == _camel(s),
  orElse: () => SyncStatus.failed,
);

String _camel(String enumWireValue) {
  // QUEUED_LOCAL -> queuedLocal
  final parts = enumWireValue.toLowerCase().split('_');
  return parts.first +
      parts.skip(1).map((w) => w[0].toUpperCase() + w.substring(1)).join();
}

String _wire(SyncStatus s) {
  switch (s) {
    case SyncStatus.queuedLocal:
      return 'QUEUED_LOCAL';
    case SyncStatus.syncing:
      return 'SYNCING';
    case SyncStatus.synced:
      return 'SYNCED';
    case SyncStatus.conflict:
      return 'CONFLICT';
    case SyncStatus.failed:
      return 'FAILED';
  }
}

class QueuedOp {
  QueuedOp({
    required this.opUuid,
    required this.entityType,
    this.entityRef,
    required this.action,
    required this.payload,
    required this.clientTs,
    required this.status,
  });

  final String opUuid;
  final String entityType;
  final String? entityRef;
  final String action;
  final Map<String, dynamic> payload;
  final String clientTs;
  final SyncStatus status;
}

/// Local write queue: farmer app writes here FIRST with no signal, then
/// flushes to POST /api/v1/sync when connectivity returns. Mirrors
/// app/lib/offlineQueue.ts's OfflineQueue class 1:1, backed by drift/SQLite
/// instead of AsyncStorage/JSON blob, and the server is the same idempotent
/// (op_uuid), server-wins offlineSyncService.
class OfflineQueue {
  OfflineQueue(this._db, this._api);

  final OfflineQueueDb _db;
  final ApiClient _api;
  static const _uuid = Uuid();

  /// Enqueue a write locally. Works with no signal — never throws on
  /// network failure since it never touches the network.
  Future<QueuedOp> enqueue({
    required String entityType,
    String? entityRef,
    required String action,
    required Map<String, dynamic> payload,
  }) async {
    final op = QueueOpsCompanion.insert(
      opUuid: _uuid.v4(),
      entityType: entityType,
      entityRef: Value(entityRef),
      action: action,
      payloadJson: jsonEncode(payload),
      clientTs: DateTime.now().toIso8601String(),
      status: _wire(SyncStatus.queuedLocal),
    );
    await _db.into(_db.queueOps).insert(op);
    return QueuedOp(
      opUuid: op.opUuid.value,
      entityType: entityType,
      entityRef: entityRef,
      action: action,
      payload: payload,
      clientTs: op.clientTs.value,
      status: SyncStatus.queuedLocal,
    );
  }

  Future<List<QueueOp>> pending() =>
      (_db.select(_db.queueOps)..where(
            (t) => t.status.isIn([
              _wire(SyncStatus.queuedLocal),
              _wire(SyncStatus.failed),
            ]),
          ))
          .get();

  /// Flush pending ops to the server. Safe to call repeatedly (idempotent
  /// via op_uuid on the server). Conflicts are surfaced (never silently
  /// overwritten) so callers can notify the farmer per CLAUDE.md's
  /// "server-wins + farmer notify" offline queue state machine.
  Future<List<String>> flush() async {
    final pendingOps = await pending();
    if (pendingOps.isEmpty) return [];

    for (final op in pendingOps) {
      await (_db.update(_db.queueOps)..where((t) => t.opUuid.equals(op.opUuid)))
          .write(QueueOpsCompanion(status: Value(_wire(SyncStatus.syncing))));
    }

    final conflictedIds = <String>[];
    try {
      final body = {
        'ops': pendingOps
            .map(
              (o) => {
                'opUuid': o.opUuid,
                'entityType': o.entityType,
                'entityRef': o.entityRef,
                'action': o.action,
                'payload': jsonDecode(o.payloadJson),
                'clientTs': o.clientTs,
              },
            )
            .toList(),
      };
      final res = await _api.post('/sync', body);
      final results =
          (res is Map ? res['results'] ?? res['data'] ?? res : res) as List? ??
          [];
      final byId = {
        for (final r in results) r['opUuid'] as String: r['status'] as String,
      };

      for (final op in pendingOps) {
        final s = byId[op.opUuid];
        SyncStatus next;
        if (s == 'APPLIED' || s == 'DUPLICATE') {
          next = SyncStatus.synced;
        } else if (s == 'CONFLICT') {
          next = SyncStatus.conflict;
          conflictedIds.add(op.opUuid);
        } else {
          next = SyncStatus.failed;
        }
        await (_db.update(_db.queueOps)
              ..where((t) => t.opUuid.equals(op.opUuid)))
            .write(QueueOpsCompanion(status: Value(_wire(next))));
      }
    } catch (_) {
      // Still offline — roll back to QUEUED_LOCAL for the next attempt.
      for (final op in pendingOps) {
        await (_db.update(
          _db.queueOps,
        )..where((t) => t.opUuid.equals(op.opUuid))).write(
          QueueOpsCompanion(status: Value(_wire(SyncStatus.queuedLocal))),
        );
      }
    }
    return conflictedIds;
  }

  Stream<List<QueueOp>> watchAll() => _db.select(_db.queueOps).watch();

  SyncStatus statusOf(QueueOp row) => _statusFromString(row.status);
}

File: app/lib/offlineQueue.ts
/**
 * Offline write queue (client half of the offline-first foundation, §10).
 *
 * The farmer app writes logbook entries / receipts / evidence here FIRST — with
 * no signal — then flushes to POST /api/v1/sync when connectivity returns. The
 * server (shared/services/offlineSyncService) is idempotent (op_uuid) and
 * resolves conflicts server-wins, returning a CONFLICT status the app surfaces
 * to the farmer.
 *
 * State machine (mirrors the server): QUEUED_LOCAL → SYNCING → SYNCED | CONFLICT
 *
 * Storage is injected (AsyncStorage / SQLite / MMKV) so this stays testable and
 * harness-agnostic. This is the Phase-0 foundation; wiring to real screens is
 * Phase 1+.
 */

export type SyncStatus = 'QUEUED_LOCAL' | 'SYNCING' | 'SYNCED' | 'CONFLICT' | 'FAILED';

export interface QueueOp {
  opUuid: string;              // client-generated (idempotency key)
  entityType: string;         // 'DairyCostEvent' | 'DairyRevenueEvent' | 'CoopInputOrder' | …
  entityRef?: string;         // target uuid for UPDATE
  action: 'CREATE' | 'UPDATE';
  payload: Record<string, unknown>;
  clientTs: string;           // ISO — capture time, used for server-wins ordering
  status: SyncStatus;
}

export interface KVStore {
  getItem(key: string): Promise<string | null>;
  setItem(key: string, value: string): Promise<void>;
}

const KEY = 'akcc.offline.queue.v1';

const uuid = (): string =>
  // RN crypto-lite; replace with expo-crypto in the app harness.
  'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
  });

export class OfflineQueue {
  constructor(private store: KVStore, private syncUrl: string) {}

  private async readAll(): Promise<QueueOp[]> {
    const raw = await this.store.getItem(KEY);
    return raw ? (JSON.parse(raw) as QueueOp[]) : [];
  }
  private async writeAll(ops: QueueOp[]): Promise<void> {
    await this.store.setItem(KEY, JSON.stringify(ops));
  }

  /** Enqueue a write locally. Works with no signal. */
  async enqueue(op: Omit<QueueOp, 'opUuid' | 'status' | 'clientTs'> & Partial<Pick<QueueOp, 'clientTs'>>): Promise<QueueOp> {
    const ops = await this.readAll();
    const full: QueueOp = {
      opUuid: uuid(),
      clientTs: op.clientTs ?? new Date().toISOString(),
      status: 'QUEUED_LOCAL',
      entityType: op.entityType,
      entityRef: op.entityRef,
      action: op.action,
      payload: op.payload,
    };
    ops.push(full);
    await this.writeAll(ops);
    return full;
  }

  /** Flush pending ops to the server. Safe to call repeatedly (idempotent). */
  async flush(fetchImpl: typeof fetch = fetch): Promise<QueueOp[]> {
    const ops = await this.readAll();
    const pending = ops.filter((o) => o.status === 'QUEUED_LOCAL' || o.status === 'FAILED');
    if (pending.length === 0) return ops;

    pending.forEach((o) => (o.status = 'SYNCING'));
    await this.writeAll(ops);

    try {
      const res = await fetchImpl(this.syncUrl, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ ops: pending }),
      });
      const results = (await res.json()) as Array<{ opUuid: string; status: string }>;
      const byId = new Map(results.map((r) => [r.opUuid, r.status]));
      for (const o of ops) {
        const s = byId.get(o.opUuid);
        if (s === 'APPLIED' || s === 'DUPLICATE') o.status = 'SYNCED';
        else if (s === 'CONFLICT') o.status = 'CONFLICT'; // surface to farmer
        else if (s === 'FAILED') o.status = 'FAILED';
      }
    } catch {
      // Still offline — roll back to QUEUED_LOCAL for the next attempt.
      pending.forEach((o) => (o.status = 'QUEUED_LOCAL'));
    }
    await this.writeAll(ops);
    return ops;
  }
}
